linfa-clustering/src/gaussian_mixture/algorithm.rs

```diff
@@ -0,0 +1,494 @@
+        observations: &ArrayBase<D, Ix2>,
+        resp: &Array2<F>,
+        _covar_type: &GmmCovarType,
+        reg_covar: F,
+    ) -> (Array1<F>, Array2<F>, Array3<F>) {
+        let nk = resp.sum_axis(Axis(0)) + F::from(10.).unwrap() * F::epsilon();
+        let nk2 = nk.to_owned().insert_axis(Axis(1));
+        let means = resp.t().dot(observations) / nk2;
+        // GmmCovarType = Full
+        let covariances =
+            Self::estimate_gaussian_covariances_full(&observations, resp, &nk, &means, reg_covar);
+        (nk, means, covariances)
+    }
+
+    fn estimate_gaussian_covariances_full<D: Data<Elem = F>>(
+        observations: &ArrayBase<D, Ix2>,
+        resp: &Array2<F>,
+        nk: &Array1<F>,
+        means: &Array2<F>,
+        reg_covar: F,
+    ) -> Array3<F> {
+        let n_clusters = means.nrows();
+        let n_features = means.ncols();
+        let mut covariances = Array::zeros((n_clusters, n_features, n_features));
+        for k in 0..n_clusters {
+            let diff = observations - &means.slice(s![k..k + 1, ..]);
+            let m = diff.t().to_owned() * resp.slice(s![.., k]);
+            let mut cov_k = m.dot(&diff) / nk[k];
+            let diag = cov_k.diag().to_owned() + reg_covar;
+            cov_k.diag_mut().assign(&diag);
+            covariances.slice_mut(s![k, .., ..]).assign(&cov_k);
+        }
+        covariances
+    }
+
+    fn compute_precision_cholesky_full<D: Data<Elem = F>>(
+        covariances: &ArrayBase<D, Ix3>,
+    ) -> Array3<F> {
+        let n_clusters = covariances.shape()[0];
+        let n_features = covariances.shape()[1];
+        let mut precisions_chol = Array::zeros((n_clusters, n_features, n_features));
+        for (k, covariance) in covariances.outer_iter().enumerate() {
+            let cov: Array2<f64> = covariance.mapv(|v| v.into());
+            match cov.cholesky(UPLO::Lower) {
+                Ok(cov_chol) => {
+                    let sol = cov_chol
+                        .solve_triangular(UPLO::Lower, Diag::NonUnit, &Array::eye(n_features))
+                        .unwrap()
+                        .to_owned();
+                    precisions_chol.slice_mut(s![k, .., ..]).assign(&sol.t());
+                }
+                Err(_) => panic!(
+                    "Fitting the mixture model failed because some components have \
+                ill-defined empirical covariance (for instance caused by singleton \
+                or collapsed samples). Try to decrease the number of components, \
+                or increase reg_covar."
+                ),
+            };
+        }
+        precisions_chol.mapv(|v| F::from(v).unwrap())
+    }
+
+    fn e_step<D: Data<Elem = F>>(&self, observations: &ArrayBase<D, Ix2>) -> (F, Array2<F>) {
+        let (log_prob_norm, log_resp) = self.estimate_log_prob_resp(&observations);
+        let log_mean = log_prob_norm.sum() / F::from(log_prob_norm.len()).unwrap();
```

Comment:
`log_prob_norm.mean()` should also be possible

Reply:
ok